stl/inc/flat_map

```diff
@@ -0,0 +1,953 @@
+// flat_map standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef _FLAT_MAP_
+#define _FLAT_MAP_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#include <algorithm>
+#include <compare>
+#include <concepts>
+#include <cstdint>
+#include <initializer_list>
+#include <memory>
+#include <ranges>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <xmemory>
+#include <xutility>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+template <class Key, class T, class Compare = less<Key>, class KeyContainer = vector<Key>,
+    class MappedContainer = vector<T>>
+    requires same_as<Key, typename KeyContainer::value_type> && same_as<T, typename MappedContainer::value_type>
+class flat_map;
+
+struct sorted_unique_t {
+    explicit sorted_unique_t() = default;
+};
+inline constexpr sorted_unique_t sorted_unique{};
+
+template <class _Alloc, class _Key_container, class _Mapped_container>
+concept _Valid_Allocator_for_flat_map =
+    uses_allocator_v<_Key_container, _Alloc> && uses_allocator_v<_Mapped_container, _Alloc>;
+
+template <class _Compare>
+concept _Compare_is_transparent = requires { typename _Compare::is_transparent; };
+
+template <class _Key, class _T, class _Key_compare>
+struct _Flat_value_compare {
+    struct value_compare {
+    public:
+        bool operator()(pair<const _Key&, const _T&> _X, pair<const _Key&, const _T&> _Y) const {
+            return _Key_compare_for_val(_X.first, _Y.first);
+        }
+
+        value_compare(_Key_compare _Comp) : _Key_compare_for_val(_Comp) {}
+
+    private:
+        _Key_compare _Key_compare_for_val;
+    };
+};
+
+template <class _Key_container, class _Mapped_container>
+struct _Flat_Container {
+    struct container {
+        _Key_container keys;
+        _Mapped_container values;
+    };
+};
+
+template <class _Ty>
+struct _NODISCARD _Clear_flat_map_scope_guard {
+    _Ty* _Clearable;
+    _Clear_flat_map_scope_guard(_Ty* _Clearable) : _Clearable(_Clearable) {}
+
+    ~_Clear_flat_map_scope_guard() {
+        if (_Clearable) {
+            _Clearable->clear();
+        }
+    }
+};
+
+// Implementation
+
+template <class _KeyContainer, class _MappedContainer, bool _IsConst>
+class _Flat_map_iterator_Impl {
+public:
+    using _Key_iterator_t = typename _KeyContainer::const_iterator;

```

Comment:
I assume the "typename" can be removed, please check https://github.com/microsoft/STL/issues/3718